docs/dev/terraform.md

```diff
@@ -0,0 +1,108 @@
+extract only the needed providers.
+
+For example, the "bootstrap" gcp stage requires the google and ignition terraform providers. The installer will extract
+the terraform binary to ${stage-dir}/bin/terraform. The installer will extract the google and ignition terraform
+providers to ${stage-dir}/plugins. The terraform exec is configured to use ${stage-dir}/plugins as the explicit
+directory for the plugins.
+
+## Adding a new terraform provider
+
+To add a new terraform provider, create a directory under ./terraform. For example, if you want to add a terraform
+provider named mycloud, then you would create the ./terraform/mycloud directory.
+
+### Public terraform provider
+
+If the terraform provider you want to add is a public terraform provider, create a tools.go file with an unnamed import
+to the main package of the terraform provider. Add a go.mod file with a require statement to the terraform provider
+module.
+
+For example, let's say that you are adding the mycloud terraform provider. The mycloud terraform provider is in the
+github.com/hashicorp/terraform-provider-mycloud repository. The main package of the terraform provider is
+github.com/hashicorp/terraform-provider-mycloud. The version of the terraform provider is v1.2.3.
+
+```go
+// ./terraform/mycloud/tools.go
+package main
+
+import (
+	_ "github.com/hashicorp/terraform-provider-mycloud"
+)
+```
+
+```go
+// ./terraform/mycloud/go.mod
+module aws
+
+go 1.17
+
+require github.com/hashicorp/terraform-provider-mycloud v1.2.3
+```
+
+After creating the directory and adding the tools.go and go.mod file, run the following commands in the mycloud
+sub-pacakge.
+1. `go mod tidy`
+2. `go mod vendor`
+
+When there is a tools.go file, the Makefile will build the terraform provider from the package referenced in the unnamed
+import in the tools.go file. For the example mycloud terraform provider, the terraform provider will be built from the
+./terraform/mycloud/vendor/github.com/hashicorp/terraform/provider-mycloud directory.
+
+### Custom terraform provider
+
+If the terraform provider you want to add is a custom terraform provider, then create a main.go file for the custom
+terraform provider.
+
+When there is a main.go file, the Makefile will build the terraform provider from the top-level package of the custom
+terraform provider.
+
+## Update the version of terraform or a provider
+
+To update the version of terraform or a provider, modify the version in the go.mod file for the relevant sub-package.
+To update the version of terraform, change the version of the required module in ./terraform/terraform/go.mod. To update
+the version of the aws terraform provider, change the version of the required module in ./terraform/aws/go.mod.
+
+After updating the require statement in the go.mod file, perform the following actions in the sub-package.
+1. Remove all of the replace statements in the go.mod file.
```

Comment:
I don't understand the context/significance of this step. Can you spell it out a bit? My confusion is: if there were replaces I would expect them to be there for a reason...

Also I spot checked the go.mod in a handful of providers and the only replace statement I saw was in vsphere for our fork. I'm not sure how that replace fits in with this step.

Reply:
Ah, this is a mistake on my part. I meant to say remove all of the indirect requires. I will fix it up.

As for replaces, if there are any, they should match the replaces from the terraform provider. For example, the ibm provider has replaces for github.com/softlayer/softlayer-go and github.com/dgrijalva/jwt-go.

Reply:
Fix in 28ff94e5f...bfa3cd984.